### PR TITLE
Wrap blaze EOF in a SocketException

### PR DESCRIPTION
**Pull request by rossabaker**
Fixes secondary complaint in http4s/blaze#292, as well as a colleague's Slack lamentation.

It sucks to get an `EOF` without any context about the request that failed.  If an `EOF` bubbles up from blaze, this wraps it in a `SocketException`.  The exception mentions the request key in the message and comes with a trace.

This is a semantically breaking change: anybody who themselves was catching `EOF` directly will be in for a surprise.  We will need to trumpet this change loudly if we proceed.

### CODE REVIEW

File: blaze-client/src/main/scala/org/http4s/blaze/client/BlazeClient.scala [hunk 1]
```diff
@@ -85,6 +86,7 @@ object BlazeClient {
           borrow.use { next =>
             val res: F[Resource[F, Response[F]]] = next.connection
               .runRequest(req)
+              .adaptError { case EOF => new ResponseException(req) }
```
Comment (rossabaker) [inline blaze-client/src/main/scala/org/http4s/blaze/client/BlazeClient.scala:89]: Other exceptions can be raised by Blaze, and they're worth contemplating.  We explicitly test that a `ConnectException` percolates up.  That has request info in its message, but does not return the request for further introspection.  Should we wrap that too?  Discuss.

File: core/src/main/scala/org/http4s/HttpException.scala [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.http4s
+
+import java.io.IOException
+
+/** Indicates a general failure communicating over HTTP */
+class HttpException(final val message: String = "", final val cause: Option[Throwable] = None)
```
Comment (rossabaker) [inline core/src/main/scala/org/http4s/HttpException.scala:22]: This is speculative and maybe should go away.  I'm thinking of problems that happen before we have a request to operate on.

File: core/src/main/scala/org/http4s/HttpException.scala [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+    extends IOException(message) {
+  cause.foreach(initCause)
+}
+
+/** Indicates a failure to obtain a response to a request */
+class ResponseException[F[_]](
+    final val request: Request[F],
```
Comment (rossabaker) [inline core/src/main/scala/org/http4s/HttpException.scala:29]: Consuming this request body again is probably dangerous, which might make this an outstanding use case for our new `RequestPrelude`.

File: blaze-client/src/test/scala/org/http4s/blaze/client/BlazeClientSuite.scala [hunk 4]
```diff
@@ -244,4 +246,25 @@ class BlazeClientSuite extends BlazeClientBase {
       }
       .assert
   }
+
+  test("Blaze HTTP/1 client should raise a ResponseException when it receives an unexpected EOF") {
+    SocketGroup[IO](testBlocker).use {
+      _.serverResource[IO](new InetSocketAddress(0))
+        .map { case (addr, sockets) =>
+          val uri = Uri.fromString(s"http://[${addr.getHostName}]:${addr.getPort}/eof").yolo
```
Comment (rossabaker) [inline blaze-client/src/test/scala/org/http4s/blaze/client/BlazeClientSuite.scala:254]: Our URI is a damn disgrace.  This test will be flaky if an IPv6 address isn't returned.

File: core/src/main/scala/org/http4s/HttpException.scala (hunk 3 above)
Comment (rossabaker) [reply core/src/main/scala/org/http4s/HttpException.scala:29]: Or perhaps we don't need a custom `ResponseException` at all.  Wrapping in a standard `IOException` would get a trace, let us describe the error, and anyone who wants more from the request can always create a client middleware.

File: core/src/main/scala/org/http4s/HttpException.scala (hunk 2 above)
Comment (rossabaker) [reply core/src/main/scala/org/http4s/HttpException.scala]: Gone.

File: core/src/main/scala/org/http4s/HttpException.scala (hunk 3 above)
Comment (rossabaker) [reply core/src/main/scala/org/http4s/HttpException.scala]: Went this direction with it.

File: blaze-client/src/main/scala/org/http4s/blaze/client/BlazeClient.scala [hunk 5]
```diff
@@ -85,6 +87,9 @@ object BlazeClient {
           borrow.use { next =>
             val res: F[Resource[F, Response[F]]] = next.connection
               .runRequest(req)
+              .adaptError { case EOF =>
+                new SocketException(s"HTTP connection closed: ${key}")
```
Comment (rossabaker) [inline blaze-client/src/main/scala/org/http4s/blaze/client/BlazeClient.scala:91]: It's unusual in the JVM to log the host info with the `IOException`s, which has me wondering if there's a reason for that.  But it's consistent with what blaze does on connection failures.

**Comment by rossabaker** [comment]
The trace we get in the unit test doesn't point to the client call site.  Maybe I'm doing it wrong.

**Comment by rossabaker** [comment]
If we're okay with the new exception, this would be nice to get into the next release.

**Review by ChristopherDavenport — APPROVED** [review]

**Comment by rossabaker** [comment]
We'll do two release trains this week so I can stabilize the GHSA merges.

**Review by isomarcte — APPROVED** [review]

**Pull request merged**